src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -67,434 +113,658 @@ public JsonWebToken(string header, string payload)
-        private void Decode(string[] tokenParts, string rawData)
+        internal byte[] MessageBytes{ get; set; }
+
+        internal int NumberOfDots { get; set; }
+
+        private void ReadToken(string encodedJson)
         {
-            try
+            // JWT must have 2 dots
+            Dot1 = encodedJson.IndexOf('.');
+            if (Dot1 == -1)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+            if (Dot1 == encodedJson.Length)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+            Dot2 = encodedJson.IndexOf('.', Dot1 + 1);
+            if (Dot2 == -1)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+            if (Dot2 == encodedJson.Length)
+                Dot3 = -1;
+            else
+                Dot3 = encodedJson.IndexOf('.', Dot2 + 1);
+
+            // JWS has two dots
+            if (Dot3 == -1)
             {
-                Header = JObject.Parse(Base64UrlEncoder.Decode(tokenParts[0]));
+                IsSigned = !(Dot2 + 1 == encodedJson.Length);
+                try
+                {
+#if NET45
+                    _sChars = IsSigned ? encodedJson.ToCharArray(Dot2 + 1, encodedJson.Length - Dot2 - 1) : string.Empty.ToCharArray();
+                    SignatureBytes = Base64UrlEncoder.UnsafeDecode(_sChars);
+                    _hChars = encodedJson.ToCharArray(0, Dot1);
+                    Header = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_hChars));
+#else
+                    Header = new JsonClaimSet(JwtTokenUtilities.GetJsonDocumentFromBase64UrlEncodedString(encodedJson, 0, Dot1));
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, encodedJson.Substring(0, Dot1), encodedJson), ex));
+                }
+
+                try
+                {
+#if NET45
+                    MessageBytes = Encoding.UTF8.GetBytes(encodedJson.ToCharArray(0, Dot2));
+                    _pChars = encodedJson.ToCharArray(Dot1 + 1, Dot2 - Dot1 - 1);
+                    Payload = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_pChars));
+#else
+                    Payload = new JsonClaimSet(JwtTokenUtilities.GetJsonDocumentFromBase64UrlEncodedString(encodedJson, Dot1 + 1, Dot2 - Dot1 - 1));
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14101, encodedJson.Substring(Dot2, Dot2 - Dot1), encodedJson), ex));
+                }
             }
-            catch (Exception ex)
+            else
```

Comment:
>else [](http://example.com/codeflow?start=12&length=4)

This implies a JWE right? Might be helpful to have a comment outlining the format so people don't have to leave for the spec

>      BASE64URL(UTF8(JWE Protected Header)) || '.' ||
>      BASE64URL(JWE Encrypted Key) || '.' ||
>      BASE64URL(JWE Initialization Vector) || '.' ||
>      BASE64URL(JWE Ciphertext) || '.' ||
>      BASE64URL(JWE Authentication Tag)

Reply:
@TimHannMSFT i think in order to understand how this all works, one needs to read the spec.